### PR TITLE
templated VectorXd arguments in RBM

### LINKED ISSUE #731: c++ doKinematics, HandC, etc should take VectorXd's (properly templated) instead of double*
in the RigidBodyManipulator class.

@avalenzu , @hongkai-dai , @tkoolen - can you think of any reason to avoid this?  I think it's just vestigial (those were the first methods we implemented).

### PR DESCRIPTION
**Pull request by psiorx**
Resolves #731

### CODE REVIEW

File: systems/plants/HandCmex.cpp [hunk 1]
```diff
@@ -29,13 +29,13 @@ void mexFunction( int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[] ) {
...
   
-  double *q,*qd;
   Map<MatrixXd> *f_ext=NULL;
   Map<MatrixXd> *df_ext=NULL;
   if (static_cast<int>(mxGetNumberOfElements(prhs[1]))!=model->num_dof || static_cast<int>(mxGetNumberOfElements(prhs[2]))!=model->num_dof)
     mexErrMsgIdAndTxt("Drake:HandCmex:BadInputs","q and qd must be size %d x 1",model->num_dof);
-  q = mxGetPr(prhs[1]);
-  qd = mxGetPr(prhs[2]);
+  Map<VectorXd> q(mxGetPr(prhs[1]), model->num_dof);
+  Map<VectorXd> qd(mxGetPr(prhs[2]), model->num_dof);
```
Comment (RussTedrake) [inline systems/plants/HandCmex.cpp]: should be model->num_velocities.

File: systems/controllers/QPControllermex.cpp [hunk 2]
```diff
@@ -190,8 +190,8 @@ void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
   
   Map< VectorXd > qddot_des(mxGetPr(prhs[narg++]),nq);
   
-  double *q = mxGetPr(prhs[narg++]);
-  double *qd = &q[nq];
+  Map<VectorXd> q(mxGetPr(prhs[narg++]), nq);
+  Map<VectorXd> qd(&q[nq], nq);
```
Comment (RussTedrake) [inline systems/controllers/QPControllermex.cpp]: should be num_velocities

File: systems/controllers/bodyMotionControlmex.cpp [hunk 3]
```diff
@@ -56,7 +56,8 @@ void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
   int nq = pdata->r->num_dof;
 
   int narg = 1;
-  double *q = mxGetPr(prhs[narg++]);
+  
+  Map<VectorXd> q(mxGetPr(prhs[narg++]), nq);
   double *qd = &q[nq];
   Map< VectorXd > qdvec(qd,nq);
```
Comment (RussTedrake) [inline systems/controllers/bodyMotionControlmex.cpp]: should be num_velocities (let's catch these now)

File: systems/controllers/pelvisMotionControlmex.cpp [hunk 4]
```diff
@@ -76,14 +76,14 @@ void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
 
   int nq = pdata->r->num_dof;
   int narg = 1;
-  double *q = mxGetPr(prhs[narg++]);
+  Map<VectorXd> q(mxGetPr(prhs[narg++]), nq);
   double *qd = &q[nq];
   Map<VectorXd> qdvec(qd,nq);
```
Comment (RussTedrake) [inline systems/controllers/pelvisMotionControlmex.cpp]: num_velocities

File: systems/controllers/supportDetectmex.cpp [hunk 5]
```diff
@@ -56,6 +56,9 @@ void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
   double *q = mxGetPr(prhs[narg++]);
   double *qd = &q[nq];
   
+  Map<VectorXd> qvec(q,nq);
+  Map<VectorXd> qdvec(qd, nq);
```
Comment (RussTedrake) [inline systems/controllers/supportDetectmex.cpp]: num_velocities

File: systems/plants/RigidBodyManipulator.h [hunk 6]
```diff
@@ -41,7 +41,12 @@ class DLLEXPORT_RBM RigidBodyManipulator
...
   void compile(void);  // call me after the model is loaded
-  void doKinematics(double* q, bool b_compute_second_derivatives=false, double* qd=NULL);
+  
+  template <typename Derived>
+  void doKinematics(MatrixBase<Derived> & q, bool b_compute_second_derivatives = false);
+
+  template <typename Derived>
+  void doKinematics(MatrixBase<Derived> & q, bool b_compute_second_derivatives, MatrixBase<Derived> & qd);
 
   void doKinematicsNew(double* q, bool compute_gradients = false, double* v = nullptr, bool compute_JdotV = false);
```
Comment (RussTedrake) [inline systems/plants/RigidBodyManipulator.h]: update this one, too?

File: systems/plants/doKinematicsmex.cpp [hunk 7]
```diff
@@ -21,14 +21,19 @@ void mexFunction( int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[] ) {
...
-  q = mxGetPr(prhs[1]);
+  
+  Map<VectorXd> q(mxGetPr(prhs[1]), nq);
   bool b_compute_second_derivatives = (mxGetScalar(prhs[2])!=0.0);
   if (mxGetNumberOfElements(prhs[3])>0) 
-    qd = mxGetPr(prhs[3]);
-     
-  model->doKinematics(q,b_compute_second_derivatives,qd);
+  {
+    Map<VectorXd> qd(mxGetPr(prhs[3]), nq);
```
Comment (RussTedrake) [inline systems/plants/doKinematicsmex.cpp]: num_velocities

File: systems/plants/getCMMmex.cpp [hunk 8]
```diff
@@ -32,17 +32,14 @@ void mexFunction( int nlhs, mxArray *plhs[],int nrhs, const mxArray *prhs[] ) {
   plhs[1] = mxCreateDoubleMatrix(6,model->num_dof,mxREAL);
   Map<MatrixXd> Adot(mxGetPr(plhs[1]),6,model->num_dof);
 
-  double* qd;
+  
   if (nrhs > 2) {
-    qd = mxGetPr(prhs[2]);
+    Map<VectorXd> qd(mxGetPr(prhs[2]), model->num_dof);
```
Comment (RussTedrake) [inline systems/plants/getCMMmex.cpp]: num_velocities

File: systems/plants/getCMMmex.cpp [hunk 9]
```diff
@@ -32,17 +32,14 @@ void mexFunction( int nlhs, mxArray *plhs[],int nrhs, const mxArray *prhs[] ) {
...
+    model->getCMM(q,qd,A,Adot);
   }
   else {
-    qd = new double[model->num_dof];
-    for (int i = 0; i < model->num_dof; i++) {
-      qd[i] = 0;
-    }
+    VectorXd zeros = VectorXd::Zero(model->num_dof);
+    Map<VectorXd> qd(zeros.data(), model->num_dof);
```
Comment (RussTedrake) [inline systems/plants/getCMMmex.cpp]: vel

**Comment by RussTedrake** [comment]
looks awesome.  my comments are in.  thanks.

**Comment by psiorx** [comment]
Good call.  Updating these and any others I find.

**Pull request merged**